onnxruntime/core/providers/nnapi/nnapi_builtin/builders/op_builder_helpers.cc

```diff
@@ -19,24 +21,44 @@
 
 Status AddNnapiTranspose(ModelBuilder& model_builder,
                          const std::string& data_input,
-                         const std::string& perm_input, const std::vector<int32_t>& perm,
-                         const std::string& output) {
-  auto& shaper(model_builder.GetShaper());
+                         const std::string& perm_input,
+                         const gsl::span<const int32_t> perm,
+                         const std::string& output,
+                         const Shape* output_shape) {
+  if (output_shape == nullptr) {
+    auto& shaper(model_builder.GetShaper());
+
+    auto calculate_transpose_output_shape = [&](const std::string& data_input,
+                                                const std::string& output,
+                                                const gsl::span<const int32_t> perm) -> Status {
+      const Shape& input_dimen = shaper[data_input];
```

Comment:
It would be good to check if the compiler can optimize away all the overhead of these single call lambas given it's only called locally. If not it's adding unnecessary binary size. Same question applies in lots of places. 

Whilst the lambda adds some context via its name, I think that could equally be done via a comment and a local block.

e.g. wouldn't this do the same without all the lambda overhead?

```c++
  auto& shaper(model_builder.GetShaper());

  // Calculate transpose output shape
  {
      const Shape& input_dimen = shaper[data_input];
      ORT_RETURN_IF_NOT(perm.size() == input_dimen.size(), "Invalid perm is given!");
      size_t size = input_dimen.size();
      Shape output_dimen(size);
      for (size_t i = 0; i < size; i++)
        output_dimen[i] = input_dimen[perm[i]];

      shaper.AddShape(output, output_dimen);
      output_shape = &shaper[output];
  }
```

Reply:
agree. replaced all the single call lambdas with a local block